agent/exec/containerd/controller.go

```diff
@@ -0,0 +1,364 @@
+}
+
+// Prepare creates a container and ensures the image is pulled.
+//
+// If the container has already be created, exec.ErrTaskPrepared is returned.
+func (r *controller) Prepare(ctx context.Context) error {
+	ctx = log.WithModule(ctx, "containerd")
+
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	//// Make sure all the networks that the task needs are created.
+	// TODO(ijc)
+	//if err := r.adapter.createNetworks(ctx); err != nil {
+	//	return err
+	//}
+
+	//// Make sure all the volumes that the task needs are created.
+	// TODO(ijc)
+	//if err := r.adapter.createVolumes(ctx); err != nil {
+	//	return err
+	//}
+
+	if r.pulled == nil {
+		// Launches a re-entrant pull operation associated with controller,
+		// dissociating the context from the caller's context. Allows pull
+		// operation to be re-entrant on calls to prepare, resuming from the
+		// same point after cancellation.
+		var pctx context.Context
+
+		r.pulled = make(chan struct{})
+		pctx, r.cancelPull = context.WithCancel(context.Background()) // TODO(stevvooe): Bind a context to the entire controller.
+
+		go func() {
+			defer close(r.pulled)
+			r.pullErr = r.adapter.pullImage(pctx)
+		}()
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-r.pulled:
+		if r.pullErr != nil {
+			// NOTE(stevvooe): We always try to pull the image to make sure we have
+			// the most up to date version. This will return an error, but we only
+			// log it. If the image truly doesn't exist, the create below will
+			// error out.
+			//
+			// This gives us some nice behavior where we use up to date versions of
+			// mutable tags, but will still run if the old image is available but a
+			// registry is down.
+			//
+			// If you don't want this behavior, lock down your image to an
+			// immutable tag or digest.
+			log.G(ctx).WithError(r.pullErr).Error("pulling image failed")
+		}
+	}
+
+	if err := r.adapter.create(ctx); err != nil {
+		if isContainerCreateNameConflict(err) {
+			if _, err := r.adapter.inspect(ctx); err != nil {
+				return err
+			}
+
+			// container is already created. success!
+			return exec.ErrTaskPrepared
+		}
+
+		return errors.Wrap(err, "create container failed")
+	}
+
+	return nil
+}
+
+// Start the container. An error will be returned if the container is already started.
+func (r *controller) Start(ctx context.Context) error {
+	ctx = log.WithModule(ctx, "containerd")
+
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	ctnr, err := r.adapter.inspect(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Detect whether the container has *ever* been started. If so, we don't
+	// issue the start.
+	//
+	// TODO(stevvooe): This is very racy. While reading inspect, another could
+	// start the process and we could end up starting it twice.
+	if ctnr.Status != container.Status_CREATED {
+		return exec.ErrTaskStarted
+	}
+
+	if err := r.adapter.start(ctx); err != nil {
+		return errors.Wrap(err, "starting container failed")
+	}
+
```

Comment:
`r.adapter.start` is synchronous, so the container will have started when it returns, however the dockerapi version isn't waiting for container start, but is waiting for container == healthy.

There is no health check support here yet. In the dockerapi case the checks are actually implemented in the engine and the swarmkit executor code simply reacts to the `health_status: healthy`/`health_status: unhealthy` events.

I'm not sure what the best model is for this `containerd`+`swarmkit` work (i.e. if health checks belong in `containerd` or `swarmkit`, my gut says the latter), but I'd like to punt on the whole thing for this PR and tackle it at some later date if that's ok.